src/runtime_src/core/common/memaccess.cpp

```diff
@@ -0,0 +1,199 @@
+  }
+
+  // If no size is supplied the end address is the ending address of the last available memory bank
+  // Update the input reference to reflect the change
+  size = (size == 0) ? available_size : size;
+}
+
+enum class OperationType {
+  read,
+  write
+};
+
+static void 
+perform_memory_action(xrt_core::device* device, xrt_core::aligned_ptr_type& buf, uint64_t start_addr, uint64_t size, OperationType action)
+{
+  std::vector<mem_bank_t> vec_banks;
+  std::vector<mem_bank_t>::iterator start_bank;
+
+  // Sanity check the address and size against the mem topology
+  validate_address_and_size(device, start_addr, size, vec_banks, start_bank);
+
+  uint64_t current_addr = start_addr;
+  size_t remaining_bytes_to_see = size;
+  size_t bytes_seen = 0;
+  // continue to read as long as there are bytes left to read or we run out of banks
+  for(auto it = start_bank; (it != vec_banks.end()) && (remaining_bytes_to_see > 0); ++it) {
+    // Validate the amount of memory the current memory bank has
+    uint64_t available_bank_size = 0;
+    if (it != start_bank) {
+      current_addr = it->m_base_address;
+      available_bank_size = it->m_size;
+    }
+    else 
+      available_bank_size = it->m_size - (current_addr - it->m_base_address);
+
+    // If the available bank size is less than the source buffer write what we are able to and move to the next bank
+    uint64_t bytes_to_edit = std::min(available_bank_size, size);
+
+    // TODO Clear all flags in cout?? Why??
+    auto guard = xrt_core::utils::ios_restore(std::cout);
+    // Update the buffer index based on how far we have read
+    void* current_buffer_location = static_cast<char *>(buf.get()) + bytes_seen;
+    switch (action) {
+      case OperationType::read:
+        device->unmgd_pread(current_buffer_location, bytes_to_edit, start_addr);
+        break;
+      case OperationType::write:
+      
+        device->unmgd_pwrite(current_buffer_location, bytes_to_edit, start_addr);
+        break;
+      default:
+        throw std::runtime_error("perform_memory_action: Invalid action");
+    }
+    remaining_bytes_to_see -= bytes_to_edit;
+    bytes_seen += bytes_to_edit;
+  }
+
+  if (remaining_bytes_to_see > 0)
+      throw std::runtime_error(boost::str(boost::format("Warning: Saw %llu bytes. Requested %llu bytes") % bytes_seen % size));
+}
+
+namespace xrt_core {
+  std::vector<char>
+  device_mem_read(device* device, uint64_t start_addr, uint64_t size)
+  {
+    // Allocate a buffer to hold the read data
+    auto buf = xrt_core::aligned_alloc(xrt_core::getpagesize(), size);
+    if (!buf)
+      throw std::runtime_error("read_banks: Failed to allocate aligned buffer");
+    std::memset(buf.get(), 0, size);
+
+    // Read from the device
+    perform_memory_action(device, buf, start_addr, size, OperationType::read);
+
+    // Format the read data into the return object
+    std::vector<char> data(size);
+    std::memcpy(data.data(), buf.get(), size);
```

Comment:
```suggestion
    std::vector data { buf.get(), buf.get() + size };
```
assuming `buf.get()` returns a `char*`.

Reply:
buf.get() returns a void* but this is pretty neat!